packages/filesystem/src/browser/file-tree/file-tree-model.ts

```diff
@@ -109,4 +111,51 @@ export class FileTreeModel extends TreeModel implements LocationService {
         return true;
     }
 
+    upload(destination: string, file: File, fileName: string) {
```

Comment:
why it cannot be so simple like that:
```ts
const uri = destination + '/' + fileName;
const encoding = 'base64';
const reader = new FileReader();
reader.onload = async e => {
    const fileContent: ArrayBuffer = reader.result;
    const content = base64.fromByteArray(new Uint8Array(fileContent));
    if (await this.fileSystem.exists(uri)) {
        const stat = await this.fileSystem.getFileStat(uri);
        this.fileSystem.setContent(stat, content, { encoding });
    } else {
        this.fileSystem.createFile(uri, { content, encoding });
    }
};
reader.readAsArrayBuffer(file);
```

@epatpol I wanted to point out to:
https://github.com/theia-ide/theia/blob/ec299e503a8ee359672a6d2a48a694da7488e29d/packages/userstorage/src/browser/user-storage-service-filesystem.ts#L72-L84

But it turned out to be bogus, you forget `await` for exists and checking whether promise is defined instead. Also there is `then(() => Promise.resovle))`, I don't get why.

Reply:
code modified,